category_theory/opposites.lean

```diff
@@ -79,6 +79,15 @@ definition op_inv : (Cᵒᵖ ⥤ Dᵒᵖ) ⥤ (C ⥤ D)ᵒᵖ :=
 
 end
 
+namespace category
+variables {C} {D : Type u₂} [𝒟 : category.{u₂ v₂} D]
+include 𝒟
+
+@[simp] lemma op_id_app (F : (C ⥤ D)ᵒᵖ) (X : C) : (𝟙 F : F ⟹ F).app X = 𝟙 (F.obj X) := rfl
```

Comment:
I have to say I'm suspicious of this simp lemma... aren't you just taking the definition of `nat_trans.id` and decorating it with the exact definitionally-equal form to match whatever you need for some specific proof to go through? I think that it has to be better to just write out the proofs explicitly, than to create this kind of one-off simp lemmas...

Reply:
Maybe this simp-lemma is bad. Maybe it becomes superfluous after we have a good `op`-interface with corresponding simp-lemmas. But I don't see how this "bad" simp-lemma would hurt (I'm uneducated in this respect). And I really don't like the idea of explicitly writing out proofs that `tidy` could do with this simp-lemma.

Reply:
It's not that the simp lemma is itself bad or problematic. Rather, I would argue that an artificial-looking lemma like this is actually part of whatever proof it was written for the benefit of. Effectively, you have written out an intermediate step in the proof after all. But the proof is now split between two files, with no explicit reference from either part of the proof to the other part. I can infer it must have to do with one of your new proofs (though I'm not entirely sure which!) because I'm looking at this commit, but a future reader of the code will have no idea.

This seems undesirable for software engineering sorts of reasons. For example, a refactor might break the link between the parts of the proof, with no indication available of how the proof was supposed to work. More likely, the main part of the proof could be refactored/modified/deleted, and we would still be stuck with the simp lemma because we would have no idea that it is now unused. You might call this setup "anti-modular", because some bit of `limits.lean` has an implicit dependency on details of `opposites.lean`. In the long run, it's better to avoid this kind of coupling.

Reply:
I guess this will all be moot anyways, since it looks like #510 will break this proof...? The simp lemma would no longer type check.

Reply:
The full proof is
```lean
def cocones : (J ⥤ C)ᵒᵖ ⥤ (C ⥤ Type v) :=
{ obj := functor.cocones,
  map := λ F G f, whisker_left _ (coyoneda.map f),
  map_id' := by intros; rw coyoneda.map_id; apply whisker_left_id,
  map_comp' := by intros; rw coyoneda.map_comp; apply whisker_left_vcomp }
```
`whisker_left_id` and `whisker_left_vcomp` are simp lemmas, but they don't get applied automatically because they are written in terms of `nat_trans.id` and `nat_trans.vcomp` and not identities and composition in the functor category.

Reply:
Right, so this is more reason to treat `nat_trans`s as morphisms everywhere... I guess I feel another small refactor PR is brewing...

Reply:
I suggest that we wait with this PR to see what happens if #510 is merged. Then we can see how to deal with this in a good way.